lib/factor/factorQuadratic.js

```diff
@@ -0,0 +1,127 @@
+// `ax^2 + bx + c`, and
+// if it is, will factor it using on of the following rules:
+//    - Difference of squares e.g. x^2 - 4 -> (x+2)(x-2)
+//    - Perfect square e.g. x^2 + 2x + 1 -> (x+1)^2
+//    - Sum/product rule e.g. x^2 + 3x + 2 -> (x+1)(x+2)
+//    - TODO: quadratic formula
+function factorQuadratic(node) {
+  node = flatten(node);
+  if (!checks.isQuadratic(node)) {
+    return Node.Status.noChange(node);
+  }
+
+  // get a, b and c
+  let firstTermNode, secondTermNode, constantTermValue;
+  for (const term of node.args) {
+    if (Node.Type.isConstant(term)) {
+      constantTermValue = term.eval();
+    }
+    else if (Node.PolynomialTerm.isPolynomialTerm(term)) {
+      const polyTerm = new Node.PolynomialTerm(term);
+      const exponent = polyTerm.getExponentNode(true);
+      if (exponent.value === '2') {
+        firstTermNode = polyTerm;
+      }
+      else if (exponent.value === '1') {
+        secondTermNode = polyTerm;
+      }
+      else {
+        return Node.Status.noChange(node);
+      }
+    }
+    else {
+      return Node.Status.noChange(node);
+    }
+  }
+
+  if (!firstTermNode || !constantTermValue) {
+    return Node.Status.noChange(node);
+  }
+
+  const symbol = firstTermNode.getSymbolNode();
+  const firstTermCoeffValue = firstTermNode.getCoeffValue();
+  const firstTermCoeffRootValue = Math.sqrt(Math.abs(firstTermCoeffValue));
+  const firstTermCoeffRootNode = Node.Creator.constant(firstTermCoeffRootValue);
+  let constantTermRootValue = Math.sqrt(Math.abs(constantTermValue));
+  let constantTermRootNode = Node.Creator.constant(constantTermRootValue);
+
+  // check if difference of squares: are a and c squares and there is no b
+  if (!secondTermNode) {
+    // must be a difference of squares
+    if (constantTermValue < 0 &&
+        firstTermCoeffRootValue % 1 === 0 &&
+        constantTermRootValue % 1 === 0) {
+      const polyTerm = Node.Creator.polynomialTerm(symbol, null, firstTermCoeffRootNode);
+      const firstParen = Node.Creator.parenthesis(
+        Node.Creator.operator('+', [polyTerm, constantTermRootNode]));
+      const secondParen = Node.Creator.parenthesis(
+        Node.Creator.operator('-', [polyTerm, constantTermRootNode]));
+
+      // create node in difference of squares form
+      const newNode = Node.Creator.operator('*', [firstParen, secondParen], true);
+      return Node.Status.nodeChanged(
+        ChangeTypes.FACTOR_DIFFERENCE_OF_SQUARES, node, newNode);
+    }
+  }
+  else {
+    // check if perfect square: are a and c squares and b = 2*sqrt(a)*sqrt(c)
+
+    // if the second term is negative, then the constant in the parens is subtracted
+    // i.e. x^2 -2x + 1 -> (x-1)^2
+    const secondTermCoeffValue = secondTermNode.getCoeffValue();
+    if (secondTermCoeffValue < 0) {
+      constantTermRootValue = constantTermRootValue * -1;
+      constantTermRootNode = Negative.negate(constantTermRootNode);
```

Comment:
alternatively, you can just make `constantTermRootNode` at the times you need it instead of at the beginning

not sure if that's better or not

I have a slight preference for this because it becomes more obvious why you need it